language-plutus-core/stdlib/Language/PlutusCore/StdLib/Type.hs

```diff
@@ -1,43 +1,181 @@
 -- | This module defines Haskell data types that simplify construction of PLC types and terms.
 
-{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes        #-}
+{-# LANGUAGE TypeFamilies      #-}
+
 module Language.PlutusCore.StdLib.Type
-    ( HoledType(..)
-    , RecursiveType(..)
-    , holedTyApp
-    , holedToRecursive
+    ( (~~>)
+    , star
+    , RecursiveType (..)
+    , makeRecursiveType
     ) where
 
-import           Language.PlutusCore.Type
-
--- | A type with a hole inside. The reason for having such a thing is that 'Wrap'
--- expects the pattern functor of a recursive type while in type signatures we use
--- actual recursive types. So we need a way to construct recursive types such that from
--- any such type we can easily extract its pattern functor as well as easily use the
--- type itself. 'RecursiveType' below allows to do that (except the pattern functor is
--- already supplied to 'Wrap'), but some types are actually type functions that receive
--- arguments and only then return a recursive type (see 'getBuiltinList' for example).
--- Thus we make a type with a hole where the hole can be filled by either 'TyFix' or
--- 'id', so this type, after all arguments are supplied (see 'holedTyApp'), can be
--- converted to the corresponding 'RecursiveType' (see 'holedToRecursive').
--- See "docs/Holed types.md" for more information.
-data HoledType tyname a = HoledType
-    { _holedTypeName :: tyname a
-    , _holedTypeCont :: (Type tyname a -> Type tyname a) -> Type tyname a
-    }
+import           Language.PlutusCore
+import           Language.PlutusCore.MkPlc
+import           PlutusPrelude
 
--- | A 'Type' that starts with a 'TyFix' (i.e. a recursive type) packaged along with a
+-- | A 'Type' that starts with a 'TyIFix' (i.e. a recursive type) packaged along with a
 -- specified 'Wrap' that allows to construct elements of this type.
-data RecursiveType tyname a = RecursiveType
-    { _recursiveWrap :: forall name. Term tyname name a -> Term tyname name a
-    , _recursiveType :: Type tyname a
+data RecursiveType ann = RecursiveType
+    { _recursiveType :: Type TyName ann
+      -- ^ This is not supposed to have duplicate names.
+    , _recursiveWrap :: [Type TyName ann] -> Term TyName Name ann -> Term TyName Name ann
+      -- ^ This produces terms with duplicate names.
     }
 
--- | Apply a 'HoledType' to a 'Type'.
-holedTyApp :: HoledType tyname () -> Type tyname () -> HoledType tyname ()
-holedTyApp (HoledType name cont) arg = HoledType name $ \hole -> TyApp () (cont hole) arg
+infixr 5 ~~>
+
+zipWithAligned :: (a -> b -> c) -> [a] -> [b] -> [c]
+zipWithAligned _ []       []       = []
+zipWithAligned f (x : xs) (y : ys) = f x y : zipWithAligned f xs ys
+zipWithAligned _ _        _        = error "Lists are not of the same length."
```

Comment:
Is this really preferable to `zipWith` with a possible length check? The error message is now less clear.

Reply:
I think you're right, I'll change that.